syncer/syncer.go

```diff
@@ -818,16 +823,14 @@ func (s *Syncer) Connect(ctx context.Context, addr string) (*Peer, error) {
 		if err != nil {
 			return
 		}
+		defer done()
+
+		if err := s.addPeer(p); err != nil {
```

Comment:
This should happen outside the goroutine now. Otherwise you still get the race.

You'll also need to move `s.tg.Add` into `runPeer`, directly after the `defer` delete of the peer. To make sure we always `runPeer` when `addPeer` returns `nil`.

Reply:
Too hasty 👍 thanks